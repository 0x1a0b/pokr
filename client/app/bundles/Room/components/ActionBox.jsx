import PropTypes from 'prop-types';
import React from 'react';
import ResultPanel from '../components/ResultPanel';
import EventEmitter from 'libs/eventEmitter';

export default class ActionBox extends React.Component {

  state = {
    buttonState: this.props.roomState
  }

  showResult = (e) => {
    this.setState({buttonState: 'open'});
    if (!Cookies.get('showTip')) {
      Cookies.set('showTip', true);
    }
    publishResult();
  }

  skipStory = () => {
    if (this.props.role === 'Moderator') {
      App.rooms.perform('set_story_point', {
        roomId: this.props.roomId,
        data: { point: 'null', story_id: this.props.storyId }
      });
    }
  }

  clearVotes = () => {
    if (this.props.role === 'Moderator') {
      App.rooms.perform('clear_votes', {
        roomId: this.props.roomId,
        data: { story_id: this.props.storyId }
      });
    }
  }

  resetActionBox = () => {
    this.setState({ buttonState: 'not-open' });
  }

  setToDrawBoard = () => {
    this.setState({ buttonState: 'draw' });
  }

  showBoard = () => {
    $('#board').html('');
    drawBoard();
  }

  resetTimer = () => {
    // if (!this.props.timerInterval) {
    //   return
    // }
    // let timer;

    // if (timer > 0) {
    //   clearInterval(POKER.timer);
    // }
    // $(".timer").show();
    // $(".timer .counter").text(POKER.timerInterval);
    // let tick = POKER.timerInterval;
    // const $icon = $(".timer .fa");

    // POKER.timer = setInterval(() => {
    //   if (tick <= 0) {
    //     if (!$icon.hasClass("warning")) {
    //       $(".timer .counter").text("Time up");
    //       $icon.attr("class", "fa warning").text("⚠️");
    //     }
    //     const $warning = $(".warning");
    //     if(!$warning.is(':animated')) {
    //       $warning.fadeToggle("fast");
    //     }
    //   } else {
    //     if (!$icon.hasClass("sandglass")) {
    //       $icon.attr("class", "fa sandglass").text("⌛").show();
    //     }
    //     tick -= 1;
    //     $(".timer .counter").text(tick);
    //   }
    // }, 1000);
  }

  disableTimer = () => {
    // $(".timer").remove();
  }

  componentDidMount = () => {
    EventEmitter.subscribe("resetActionBox", this.resetActionBox);
    // EventEmitter.subscribe("roomClosed", this.setToDrawBoard);
    // EventEmitter.subscribe("roomClosed", this.disableTimer);
    if (this.props.roomState !== "draw" && this.props.timerInterval > 0) {
      EventEmitter.subscribe("resetTimer", this.resetTimer);
      this.resetTimer();      
    }
    if (this.props.roomState === 'open') {
      showResultSection();
    }
  }

  render() {
    let onClickName;
    let buttonText;
    let buttonHtml;
    let secondButtonText;
    let secondOnClickName;
    let buttonCount = 1;

    if (this.props.role === 'Moderator') {
      if (this.state.buttonState === 'not-open') {
        onClickName = this.showResult;
        buttonText = "Flip";
      } else if (this.state.buttonState === 'open') {
        onClickName = this.skipStory;
        buttonText = "Skip it";
        secondOnClickName = this.clearVotes;
        secondButtonText = "Clear votes"
        buttonCount = 2;
      } else if (this.state.buttonState === 'draw') {
        onClickName = this.showBoard;
        buttonText = "Show board";
      }
    }
    buttonHtml = ((() => {
      if (onClickName && buttonText && buttonCount > 1) {
        return (
          <div className="btn-group btn-group-lg btn-group-justified" role="group">
            <a onClick={onClickName} className="btn btn-default btn-lg btn-info" href="javascript:;" role="button">
              {buttonText}
            </a>
            <a onClick={secondOnClickName} className="btn btn-default btn-lg btn-info" href="javascript:;" role="button">
              {secondButtonText}
            </a>
          </div>
        )
      } else if(onClickName && buttonText) {
        return (
          <div className="" role="group">
            <a onClick={onClickName} className="btn btn-default btn-lg btn-info btn-block" href="javascript:;" role="button">
              {buttonText}
            </a>
          </div>
        )
      }
    }))()

    const tip = ((() => {
      // already decided point
      if (Cookies.get('showTip') && !Cookies.get('adp') && this.props.role === 'Moderator') {
        Cookies.set('adp', true);
        return (
          <div className="container-fluid" style={{clear: 'both', width: '90%'}}>
            <div className="alert alert-success alert-dismissible" role="alert">
              <button type="button" className="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>
              <strong>Tip!</strong> Click the bar below to decide the point.
            </div>
          </div>
        )
      }
    }))()

    return (
      <div className="panel panel-default">
        <div className="panel-heading">
          Action
          <span className="timer pull-right" style={{display: 'none'}}>
            <i className="fa sandglass">⌛</i>
            &nbsp;
            <i className="counter">{this.props.timerInterval}</i>
          </span>
        </div>
        <div className="panel-body row">
          <div id="actionBox" className="row">
            {tip}
            <ResultPanel roomId={this.props.roomId} role={this.props.role} storyId={this.props.storyId} />
            <div className="openButton container-fluid">
              {buttonHtml}
            </div>
          </div>
        </div>
      </div>
    )
  }
}