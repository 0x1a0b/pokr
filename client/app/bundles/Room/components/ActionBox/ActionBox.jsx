import PropTypes from 'prop-types'
import React from 'react'
import ResultPanel from './ResultPanel/ResultPanel'
import EventEmitter from 'libs/eventEmitter'
import update from 'immutability-helper'
import RoomActions from 'libs/roomActions'
import {defaultTourColor} from 'libs/barColors'

export default class ActionBox extends React.Component {
  constructor(props){
    super(props);

    this.state = {
      roomState: this.props.roomState,
      countDown: this.props.countDown,
    }
  }

  timer = () => {
    let newState = update(this.state, {
      countDown: { $set: this.state.countDown - 1 }
    })

    this.setState(newState)

    if(this.state.countDown < 1) {
      clearInterval(this.intervalId);
    }
  }

  showResult = (e) => {
    let newState = update(this.state, {
      roomState: { $set: "open" }
    })

    this.setState(newState)

    if (!Cookies.get('showTip')) {
      Cookies.set('showTip', true)
    }
    if (this.props.role === 'Moderator' && this.state.roomState !== 'open') {
      App.rooms.perform('action', {
        roomId: this.props.roomId,
        data: 'open',
        type: 'action'
      })
    }
  }

  skipStory = () => {
    if (this.props.role === 'Moderator') {
      App.rooms.perform('set_story_point', {
        roomId: this.props.roomId,
        data: { point: 'null', story_id: this.props.storyId }
      });
    }
  }

  clearVotes = () => {
    if (this.props.role === 'Moderator') {
      App.rooms.perform('clear_votes', {
        roomId: this.props.roomId,
        data: { story_id: this.props.storyId }
      });
    }
  }

  resetActionBox = () => {
    let newState = update(this.state, {
      roomState: { $set: "not-open" },
      countDown: { $set: this.props.countDown }
    })

    this.setState(newState)

    if (this.state.roomState !== "draw" && this.props.countDown > 0) {
      this.intervalId = setInterval(this.timer, 1000)
    }
  }

  showBoard = () => {
    EventEmitter.dispatch("showBoard")
  }

  componentDidMount() {
    EventEmitter.subscribe("resetActionBox", this.resetActionBox);
    if (this.props.roomState !== "draw" && this.props.countDown > 0) {
      this.intervalId = setInterval(this.timer, 1000)
    }
    if (this.props.roomState === 'open') {
      RoomActions.showResultSection();
    }

    let text
    let title
    if ("Moderator" === this.props.role) {
      title = "Actions"
      text = "All the actions with points will be done in this panel. You can call the final point of a story by clicking on the point bar, clear votes, or skip current story."
    } else {
      title = "Votes"
      text = "You'll see the result of votes here, only the moderator can decide which is the final point."
    }

    this.props.addSteps({
      title: title,
      text: text,
      selector: '#action-box',
      position: 'top-right',
      style: defaultTourColor
    })
  }

  componentWillReceiveProps(nextProps) {
    this.state = {
      roomState: nextProps.roomState,
      countDown: nextProps.countDown,
    } 
  }

  componentWillUnmount() {
    clearInterval(this.intervalId);
  }

  render() {
    let onClickName;
    let buttonText;
    let buttonHtml;
    let secondButtonText;
    let secondOnClickName;
    let buttonCount = 1;

    if (this.props.role === 'Moderator') {
      if (this.state.roomState === 'not-open') {
        onClickName = this.showResult;
        buttonText = "Flip";
      } else if (this.state.roomState === 'open') {
        onClickName = this.skipStory;
        buttonText = "Skip it";
        secondOnClickName = this.clearVotes;
        secondButtonText = "Clear votes"
        buttonCount = 2;
      } else if (this.state.roomState === 'draw') {
        onClickName = this.showBoard;
        buttonText = "Show board";
      }
    }
    buttonHtml = ((() => {
      if (onClickName && buttonText && buttonCount > 1) {
        return (
          <div className="btn-group btn-group-lg btn-group-justified" role="group">
            <a onClick={onClickName} className="btn btn-default btn-lg btn-info" href="javascript:;" role="button">
              {buttonText}
            </a>
            <a onClick={secondOnClickName} className="btn btn-default btn-lg btn-info" href="javascript:;" role="button">
              {secondButtonText}
            </a>
          </div>
        )
      } else if (onClickName && buttonText) {
        return (
          <div className="" role="group">
            <a onClick={onClickName} className="btn btn-default btn-lg btn-info btn-block" href="javascript:;" role="button">
              {buttonText}
            </a>
          </div>
        )
      }
    }))()

    const tip = ((() => {
      // already decided point
      if (Cookies.get('showTip') && !Cookies.get('adp') && this.props.role === 'Moderator') {
        Cookies.set('adp', true);
        return (
          <div className="container-fluid" style={{clear: 'both', width: '90%'}}>
            <div className="alert alert-success alert-dismissible" role="alert">
              <button type="button" className="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>
              <strong>Tip!</strong> Click the bar below to decide the point.
            </div>
          </div>
        )
      }
    }))()

    const icon = () => {
      if ("draw" !== this.props.roomState && this.props.countDown) {
        if (this.props.countDown < 1) {
          return(
            <span className="timer pull-right">
              <i className="fa warning">⚠️</i>
              Time out
            </span>
          )
        } else {
          return(
            <span className="timer pull-right">
              <i className="fa sandglass">⌛</i>
              &nbsp;
              <i className="counter">{this.props.countDown}</i>
            </span>
          )
        }
      }
    }

    return (
      <div className="panel panel-default" id="action-box">
        <div className="panel-heading">
          Action
          {icon()}
        </div>
        <div className="panel-body row">
          <div id="actionBox" className="row">
            {tip}
            <ResultPanel roomId={this.props.roomId} role={this.props.role} storyId={this.props.storyId} />
            <div className="openButton container-fluid">
              {buttonHtml}
            </div>
          </div>
        </div>
      </div>
    )
  }
}
