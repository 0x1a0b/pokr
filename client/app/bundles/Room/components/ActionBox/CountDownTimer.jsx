import PropTypes from 'prop-types'
import React from 'react'
import update from 'immutability-helper'
import EventEmitter from 'libs/eventEmitter'

export default class CountDownTimer extends React.Component {

  static propTypes = {
    countDown: PropTypes.number
  }

  constructor(props) {
    super(props)

    this.state = {
      storyId: props.storyId,
      countDown: props.countDown
    }
  }

  componentDidMount() {
    EventEmitter.subscribe("resetActionBox", this.resetTimer)
    if (this.props.countDown > 0) {
      this.intervalId = setInterval(this.timer, 1000)
    }
  }

  resetTimer = () => {
    this.setState({
      storyId: this.props.storyId,
      countDown: this.props.countDown
    })

    if (this.props.countDown > 0) {
      if (this.intervalId) {
        clearInterval(this.intervalId)
      }
      this.intervalId = setInterval(this.timer, 1000)
    }
  }

  timer = () => {
    // debugger
    let newState = update(this.state, {
      countDown: { $set: this.state.countDown - 1 }
    })

    this.setState(newState)

    if(this.state.countDown < 1) {
      clearInterval(this.intervalId);
    }
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      storyId: nextProps.storyId,
      countDown: this.state.countDown
    })
  }

  componentWillUnmount() {
    clearInterval(this.intervalId);
  }

  render() {
    if (this.state.countDown < 1) {
      // console.log("====Warning")
      return(
        <span className="timer pull-right">
          <i className="fa warning">⚠️</i>
          Time out
        </span>
      )
    } else {
      // console.log("====Counting")
      return(
        <span className="timer pull-right">
          <i className="fa sandglass">⌛</i>
          &nbsp;
          <i className="counter">{this.state.countDown}</i>
        </span>
      )
    }
  }
}

// if (this.state.roomState !== "draw" && this.props.countDown > 0) {
//   clearInterval(this.intervalId)
//   this.intervalId = setInterval(this.timer, 1000)
// }