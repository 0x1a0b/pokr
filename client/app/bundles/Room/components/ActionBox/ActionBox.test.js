import React from 'react'
import ActionBox from './ActionBox'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme';
import sinon from 'sinon';

it('renders correctly', () => {
  const addSteps = jest.fn()
  const tree = renderer.create(
    <ActionBox
      roomId="test-room"
      storyId={1}
      roomState="open"
      role="Moderator"
      countDown={10}
      addSteps={addSteps}
      />
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test("shows the vote results when click Flip button", () => {
  global.App = { rooms: { perform: function() { } } }

  const addSteps = sinon.spy()
  const component = shallow(
    <ActionBox
      roomId="test-room"
      storyId={1}
      roomState="not-open"
      role="Moderator"
      countDown={10}
      addSteps={addSteps}
      />
  )
  component.find('.flip').simulate('click');

  expect(component.find('.openButton .btn')).toHaveLength(2);
  expect(component.find('.result-panel')).toHaveLength(1);
})