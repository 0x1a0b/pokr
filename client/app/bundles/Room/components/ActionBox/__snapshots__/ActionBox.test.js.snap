// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="panel panel-default"
  id="action-box"
>
  <div
    className="panel-heading"
  >
    Action
    <span
      className="timer pull-right"
    >
      <i
        className="fa sandglass"
      >
        ⌛
      </i>
       
      <i
        className="counter"
      >
        10
      </i>
    </span>
  </div>
  <div
    className="panel-body row"
  >
    <div
      className="row"
      id="actionBox"
    >
      <div
        className="result-panel"
      >
        <div
          className="container-fluid"
          style={
            Object {
              "clear": "both",
              "width": "100%",
            }
          }
        >
          <div
            className="row-container container-fluid"
          >
            <div />
          </div>
        </div>
      </div>
      <div
        className="openButton container-fluid"
      >
        <div
          className="btn-group btn-group-lg btn-group-justified"
          role="group"
        >
          <a
            className="btn btn-default btn-lg btn-info"
            href="javascript:;"
            onClick={[Function]}
            role="button"
          >
            Skip it
          </a>
          <a
            className="btn btn-default btn-lg btn-info"
            href="javascript:;"
            onClick={[Function]}
            role="button"
          >
            Clear votes
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
